File: backend/app/routers/datasets.py
from fastapi import APIRouter, UploadFile, Form, HTTPException
from fastapi.responses import StreamingResponse
from app.services import dataset_service

router = APIRouter(prefix="/datasets", tags=["datasets"])


# --------------------------- Upload ---------------------------
@router.post("/upload")
def upload_dataset(file: UploadFile, session_id: str = Form(...)):
    try:
        return dataset_service.upload_dataset(file, session_id)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


# --------------------------- List -----------------------------
@router.get("")
def list_datasets(session_id: str):
    return dataset_service.list_datasets(session_id)


# --------------------------- Metadata -------------------------
@router.get("/{dataset_id}")
def get_dataset(dataset_id: str, session_id: str):
    dataset = dataset_service.get_dataset(dataset_id, session_id)
    if not dataset:
        raise HTTPException(status_code=404, detail="Dataset not found")
    return dataset


@router.get("/{dataset_id}/info")
def get_file_info(dataset_id: str, session_id: str):
    info = dataset_service.get_file_info(dataset_id, session_id)
    if not info:
        raise HTTPException(status_code=404, detail="Dataset not found")
    return info


@router.get("/{dataset_id}/status")
def get_file_status(dataset_id: str, session_id: str):
    status = dataset_service.get_file_status(dataset_id, session_id)
    if not status:
        raise HTTPException(status_code=404, detail="Dataset not found")
    return {"status": status}


# --------------------------- Preview --------------------------
@router.get("/{dataset_id}/preview")
def preview_dataset(dataset_id: str, session_id: str, use_cleaned: bool = False):
    rows = dataset_service.preview_dataset(dataset_id, session_id, use_cleaned=use_cleaned)
    if rows is None:
        raise HTTPException(status_code=404, detail="Dataset not found")
    return {"preview": rows}


# --------------------------- Download -------------------------
@router.get("/{dataset_id}/download")
def download_dataset(dataset_id: str, session_id: str):
    file_bytes = dataset_service.download_dataset(dataset_id, session_id)
    if not file_bytes:
        raise HTTPException(status_code=404, detail="Dataset not found")

    return StreamingResponse(
        iter([file_bytes]),
        media_type="application/octet-stream",
        headers={"Content-Disposition": f"attachment; filename={dataset_id}.csv"}
    )


# --------------------------- Delete ---------------------------
@router.delete("/{dataset_id}")
def delete_dataset(dataset_id: str, session_id: str):
    dataset_service.delete_dataset(dataset_id, session_id)
    return {"message": "Dataset deleted"}

File: frontend/src/api/datasets.ts
import axios from "axios";

const API_BASE = "https://machine-learning-tweet-sentiment.onrender.com/datasets";

export const uploadDataset = async (file: File, sessionId: string) => {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("session_id", sessionId);

  const res = await axios.post(`${API_BASE}/upload`, formData, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return res.data; // contains dataset_id
};

export const listDatasets = async (sessionId: string) => {
  const res = await axios.get(`${API_BASE}`, { params: { session_id: sessionId } });
  return res.data;
};

export const getDatasetInfo = async (datasetId: string, sessionId: string) => {
  const res = await axios.get(`${API_BASE}/${datasetId}/info`, {
    params: { session_id: sessionId },
  });
  return res.data;
};

export const getDatasetStatus = async (datasetId: string, sessionId: string) => {
  const res = await axios.get(`${API_BASE}/${datasetId}/status`, {
    params: { session_id: sessionId },
  });
  return res.data.status;
};

export const previewDataset = async (datasetId: string, sessionId: string) => {
  const res = await axios.get(`${API_BASE}/${datasetId}/preview`, {
    params: { session_id: sessionId },
  });
  return res.data.preview;
};

export const downloadDataset = async (datasetId: string, sessionId: string) => {
  const res = await axios.get(`${API_BASE}/${datasetId}/download`, {
    params: { session_id: sessionId },
    responseType: "blob", // important for file downloads
  });
  return res.data;
};

export const deleteDataset = async (datasetId: string, sessionId: string) => {
  const res = await axios.delete(`${API_BASE}/${datasetId}`, {
    params: { session_id: sessionId },
  });
  return res.data;
};